### PR TITLE
Ryan

### PR DESCRIPTION
**Pull request by youcatastrophe**
So far I've got extremely basic functionality of the program (no bonus work added), BUT I'm still not finished. I'm really needing some outside input on how to manage rankings beyond just showing them in descending order, which does not handle ties or a anything like that. I'd rather NOT ask for help with something as simple as this, but I'm afraid I would just sit here wracking my brain all day, and I wouldn't get anything done.

### CODE REVIEW

**Comment by halfghaninne** [comment]
To be clear, is resolving ties the feature that you are needing help on?

**Comment by youcatastrophe** [comment]
Yes!

**Comment by halfghaninne** [comment]
Okay so I see in your schema that you are tracking `Avenger` Objects and their `votes`.  In order to implement the tie breaker in the project specs, you would have to track which two `Avengers` face off in a battle. How might you implement that?

**Comment by youcatastrophe** [comment]
I can think of how I'd track which two are up against each other. But it wouldn't keep track of who won between those two. Does that make sense? So basically what I'm saying is that I could see myself adding a new table (let's call it "showdowns") and it has a two columns, for now we'll call each column "hero_1" and "hero_2", but I can't (at least not at this point in my head, think of how I'd keep track of who won between these two. 

Again, I know this is easy stuff, but for some reason it's just not coming to me. Thanks.

**Comment by halfghaninne** [comment]
No need for self deprecation. It's _not_ inherently easy. And that's why we have a dedicated place to help you learn!

That table is definitely the right track. What about adding a `winner_id` column to it?

**Comment by youcatastrophe** [comment]
LOL. I thought about that right after I responded. Of course I can track that! Duh!  Thanks. I'll see what I can do!

**Comment by youcatastrophe** [comment]
Alright. I've created the tables, but now I can't for the life of me figure out what to do with them to get what I need. 😩

**Comment by halfghaninne** [comment]
Hmmm. Maybe your process of increasing an Avenger's votes looks at the result (the winner_id) of a Showdown.

**Comment by youcatastrophe** [comment]
Specifically, I'm having a hard time figuring out how I'm going to get access the id of the loser in my controller. My page asks the user to vote for just one person, obviously, and I don't know what to do to get that loser's information stored somehow.

**Comment by halfghaninne** [comment]
Hm. So specifically how to tell which two Avenger's are on the page, since it's randomized?

Want to make sure I'm understanding correctly.

**Comment by youcatastrophe** [comment]
Yes! That's exactly what I'm asking.

**Comment by youcatastrophe** [comment]
For the first time (other than that time I was confused about making those google spreadsheets... but let's forget about that), I feel absolutely clueless and lost. All of the concepts make sense to me. I just can't make this work. It's like my brain isn't working. The swimming competition was fairly easy for me (just time consuming). But I feel like I'm just not wrapping my head around what needs to be done to make any of this work. Does that make sense?

**Comment by halfghaninne** [comment]
Looks like you are passing the battling Avengers' ids from the view to the form action post "/process_vote" [here](https://github.com/youcatastrophe/web-app__binary-battle/blob/57f04409527f41483a6ff7c58ed44a59cdf213d5/app/views/homepage.erb#L8)

**Comment by youcatastrophe** [comment]
Right. But only the winner gets selected.

**Comment by halfghaninne** [comment]
Ahh. One sec.

**Comment by youcatastrophe** [comment]
Ha. No problem. I'll be patient.

**Comment by halfghaninne** [comment]
What about instantiating and saving a Showdown Object/record [where you do you have access](https://github.com/youcatastrophe/web-app__binary-battle/blob/57f04409527f41483a6ff7c58ed44a59cdf213d5/app/controllers/avengers_controller.rb#L1-L4) to those two Avenger Objects?

**Comment by youcatastrophe** [comment]
Ok. How about [this](https://github.com/youcatastrophe/web-app__binary-battle/commit/5d1c3ca4f2c2cbdade79a8f4df25cf174f138088)

**Comment by halfghaninne** [comment]
Looks good so far, though I'd make sure you `.save` it! The next challenge will be making sure that the winner of the Showdown is saved with the Showdown Object you just created in the **get "/"** controller action

**Comment by youcatastrophe** [comment]
Wait... How is that possible? I thought "get" controllers were not to be used for adding/changing data? If this is a get request, shouldn't I not be doing that?

**Comment by youcatastrophe** [comment]
Whoops. I think I misread your comment, Alex. I think I see what you're saying now. I read it as if you were you saying that the winner was saved "at the same time" as the rest of the object in the "get"  request.

**Comment by halfghaninne** [comment]
Oh, yeah, that was a misreading!

**Comment by youcatastrophe** [comment]
Ok. Before I forget, thanks so much or being so helpful! I'm not used to having to go back and forth like this - and, truth be told, I was dreading it because I thought my questions would get annoying or something - it's really helping me out! Thanks for the great attitude!

**Comment by youcatastrophe** [comment]
Anyway, I'm still not sure how I'm going to save the winner and make sure it gets associated with the object, as created above in my get "/" controller.

**Comment by youcatastrophe** [comment]
I'm starting to wonder if I even need a "votes" column in my "avengers" table. Maybe that table only needs the name column? I suppose I can find a method that will count the occurrences of each time the winner id shows up in the winner column? Hmmmmm.

**Comment by halfghaninne** [comment]
Know that your flow goes something like this:

**get "/"**
- define @hero_1 and @hero_2 from random selection
- create a new Showdown Object with `@hero_1.id` and `@hero_2.id` as column values; save it
- load form

Here, you have access to:
- @hero_1
- @hero_2
- @showdown

**In the homepage form view**

Here, you have access to:
- @hero_1
- @hero_2
- @showdown

The flow:
- The form selects either `@hero_1` or `@hero_2` and passes their `id` to :

**post "/process_form"**
- this should save the `id` of the selected hero to @showdown BUT

Here we only have access to:
- the selected hero's `id`

**Comment by halfghaninne** [comment]
I'm stepping over to work with another student! Feel free to post follow ups as a PR link in Slack so that Sumeet has a chance to jump in if I'm unavailable.

**Comment by youcatastrophe** [comment]
Alright. No problem! Thanks so much for your help thus far, @halfghaninne

**Comment by youcatastrophe** [comment]
Does something like [this](https://github.com/youcatastrophe/web-app__binary-battle/commit/880f72991d95a54244f01af5fdf44320033609ae) work?

**Comment by youcatastrophe** [comment]
Right now, I'm having a great deal of difficulty at the algorithmic level regarding the feature to display to the user the winner based on matches _between_ two specific users if a tie has been determined. This is something I've thought about and worked through in my head and cannot come to a solution that seem to be viable.

**Comment by halfghaninne** [comment]
I see that you've started playing around with customized methods for this, which I think is a good inclination.

In plain English, what do you want the method to accomplish?

**Comment by youcatastrophe** [comment]
I need the method to be able to determine _if_ two  “heroes”  (in my case) have an equal vote_count (which is determined by another method (called vote_count). Once this has been confirmed (that there are two heroes with the same vote_count) I need to be able to make the method reference all of the “showdowns” that took place between only those two heroes and determine the winner as being the hero who won the most in those particular showdowns. 

That’s about as clear as I can describe it at this point. Any thoughts on how I can implement this?

**Comment by halfghaninne** [comment]
That makes sense to me. 

Where do you envision this feature coming into play? On the admin page? Results? Somewhere else?

**Comment by youcatastrophe** [comment]
I guess I envisioned it being displayed on the results page that the user gets after they have voted. Again, this "view" would only be displayed _IF_ there's a tie. Otherwise, I have a very basic "ranking" that is displayed for each hero. I know that I always like to see results immediately after I vote for something online. But I can change this, if need be.

**Comment by halfghaninne** [comment]
So would it make most sense for this `tie_breaker` method come into place in the view, or within you `rank` class method?

**Comment by youcatastrophe** [comment]
Great question! That is one I've thought about numerous times now. To be honest, I'm not sure if this is the best, but my thought is to make a new method for this that uses the "rank" class method. But maybe that's not a good idea. I don't know.

**Comment by halfghaninne** [comment]
Ah. So `Avenger.rank` would be called within this new method?

**Comment by youcatastrophe** [comment]
Yes. That's my thought, anyway.

**Comment by halfghaninne** [comment]
I like that idea!

**Comment by youcatastrophe** [comment]
Actually, I'm a little confused about how I would do that. Going back to my "rank" method, I am recalling that all it does is return a hash. Hashes are still not a strongpoint for me and I'm not sure what I can do without looking up every available method for hashes on http://docs.ruby-lang.org/en/2.0.0/Hash.html, which I can do, but it would take a lot of time.

**Comment by halfghaninne** [comment]
So in this new method, `tie_breaker`, you can call `Avenger.rank`...

IF there are two values in a row that are the same:
you compare Showdown `winner_id`s between the avenger referenced in the keys.

I think the question then becomes, how do you "preference" the key/value pair in the hash that is associated with the avenger with the most wins?

**Comment by youcatastrophe** [comment]
Yes. That pretty much sums it up!

**Comment by youcatastrophe** [comment]
Do you recommend I do some googling on "preference"-ing key/value pairs on hashes?

**Comment by halfghaninne** [comment]
Not necessarily. _Maybe_ google "reordering hash Ruby". Before that, I would also think through what a custom function to reorder the list might look like, and also use classmates as resources.

**Comment by youcatastrophe** [comment]
Just to make sure we're on the same page here, I thought that the "rank" method was just that: a custom function to reorder the list. I mean, that's what this method does. It takes the name and vote_counts and puts them into a hash and reorders them (ascending), and then I have the hash returned with the .reverse method to get it descending instead. Are you talking about something different than this method?

All the "rank" method does is put an order to the listing of my names and vote_counts. Perhaps this is a very long and roundabout way to do it.

**Comment by halfghaninne** [comment]
I meant another method that reorders the list in favor of the Avenger that has won the most Showdowns of a tied twosome.  But I may have gone too far -- for now maybe that logic lives in the `tie_breaker` method, to be refactored later.

**Comment by youcatastrophe** [comment]
Ah. I see. Thanks for the clarification. That said, I'm still not getting anywhere with this method for some reason. I've at least determined that I should _first_ create a method that finds ties before I do anything else. And even that alone is proving to be very challenging. I've checked all sorts of resources online and have found nothing that will apply to this situation. Any thoughts? 

To explain (so that you whether or not I have the right idea), I'm truing to create a method that looks at the values of the entire "rank" hash and identifies duplicates. If duplicates are found, I'd like it to be able to return the keys and values of the two duplicates. I'm not finding any way to do something like this, nor can I think of an algorithm on my own that will do this. Perhaps I'm making this too complicated? I'm all ears.

**Comment by halfghaninne** [comment]
Well it seems like you will only be using your tie breaker if you have two or more keys with the same value in your result hash from `Avenger.rank`. How can you find that information from a hash?

**Comment by youcatastrophe** [comment]
Ha. I'm not sure. That's what I've been trying to get at. I have looked and looked (and thought and thought) and cannot figure out a way to identify if there are two ore more keys with the same value. I've checked all sorts of resources online and haven't found anything that is helpful for this situation. I'm not looking for an answer here, but even being pointed in the right direction would be helpful. Thanks.

**Comment by youcatastrophe** [comment]
Right now, my best guess is to use something like the values method to return an array of arrays containing only the values. I would then need to find a way to iterate though the array of arrays and find duplicates. This seems doable, but convoluted. I can try to make this happen, I just don't know if I'm making this more difficult than it needs to be.

**Comment by sumeetjain** [comment]
Ryan, can you post an example of what the hash looks like? I want to make sure I don't offer guidance that isn't true to your need.

**Comment by youcatastrophe** [comment]
No problem.  Do you mean an example of what a returned hash would actually look like? I’m gong to assume that’s what you mean and go from there. 

Here is an example taken directly from binding.pry that was executed right at the end of my "rank" class method. 

=> [["Thor", 9],
 ["Captain America", 7],
 ["Iron Man", 4],
 ["Hawkeye", 3],
 ["Hulk", 2],
 ["Black Widow", 1]

But, it would look more like this in regular hash syntax:
=>
 {"Thor" => 9 "Captain America" => 7, "Iron Man" => 4, "Hawkeye" => 3, "Hulk", => 2, "Black Widow" =>1}

Is this what you're wanting? I just want to make sure I'm on the right page to begin with.

**Comment by sumeetjain** [comment]
Gotcha. And your goal is to group the elements in that hash by their value, because you want to know if there are any ties–right?

Let me know if [this](http://stackoverflow.com/questions/18841698/group-hash-by-values-in-ruby) helps.

**Comment by sumeetjain** [comment]
(P.S. `[["Thor", 9], ["Captain America", 7], ["Iron Man", 4], ["Hawkeye", 3], ["Hulk", 2], ["Black Widow", 1]].to_h` is a nice trick–in case you didn't have that nested array as a hash already.)

**Comment by sumeetjain** [comment]
```
[22] pry(main)> x
=> {"Thor"=>9, "Captain America"=>7, "Iron Man"=>9, "Hawkeye"=>3, "Hulk"=>2, "Black Widow"=>1}
[23] pry(main)> x.group_by{|k,v| v}
=> {9=>[["Thor", 9], ["Iron Man", 9]],
 7=>[["Captain America", 7]],
 3=>[["Hawkeye", 3]],
 2=>[["Hulk", 2]],
 1=>[["Black Widow", 1]]}
[24] pry(main)>
```

```
[15] pry(main)> a
=> {9=>[["Thor", 9], ["Iron Man", 9]],
 7=>[["Captain America", 7]],
 3=>[["Hawkeye", 3]],
 2=>[["Hulk", 2]],
 1=>[["Black Widow", 1]]}
[16] pry(main)> tied_legends = []
=> []
[17] pry(main)> a.each do |k, v|
[17] pry(main)*   if v.length > 1
[17] pry(main)*     tied_legends << v
[17] pry(main)*   end
[17] pry(main)* end
=> {9=>[["Thor", 9], ["Iron Man", 9]],
 7=>[["Captain America", 7]],
 3=>[["Hawkeye", 3]],
 2=>[["Hulk", 2]],
 1=>[["Black Widow", 1]]}
[18] pry(main)> tied_legends
=> [[["Thor", 9], ["Iron Man", 9]]]
[19] pry(main)> tied_legends[0]
=> [["Thor", 9], ["Iron Man", 9]]
[20] pry(main)> tied_legends[0][0]
=> ["Thor", 9]
[21] pry(main)> tied_legends[0][1]
=> ["Iron Man", 9]
[22] pry(main)>
```

**Comment by youcatastrophe** [comment]
Just finished writing find_ties method for Avenger class. Still a little unsure as to how to define the method. Can I make it a "self.find_ties" method? Does this method even belong in the model? I don't even want to try the method out until I can understand a few more things.

**Comment by youcatastrophe** [comment]
Basic functionality is done, unless I've missed something. Have been focusing on a numer of complex methods to help with tie-breakers, but I will give them a break (no pun intended) for now.
